File: training-data/prolog/power_function.pl
my_power(_, 0, 1) :- !.
my_power(Base, Exp, Result) :-
    Exp mod 2 =:= 0, !,
    Exp1 is Exp // 2,
    my_power(Base, Exp1, Half),
    Result is Half * Half.
my_power(Base, Exp, Result) :-
    Exp1 is Exp - 1,
    my_power(Base, Exp1, Sub),
    Result is Base * Sub.

:- my_power(2, 10, R), writeln(R).
:- my_power(3, 5, R), writeln(R).

File: training-data/zig/power_function.zig
const std = @import("std");

fn myPower(base: i32, exp: i32) i32 {
    if (exp == 0) return 1;
    if (@rem(exp, 2) == 0) {
        const half = myPower(base, @divTrunc(exp, 2));
        return half * half;
    }
    return base * myPower(base, exp - 1);
}

pub fn main() void {
    std.debug.print("{d}\n", .{myPower(2, 10)});
    std.debug.print("{d}\n", .{myPower(3, 5)});
}
